kong/tools/wrpc.lua

```diff
@@ -0,0 +1,748 @@
+end
+
+
+--- Part of wrpc_peer:call()
+--- If calling the same method with the same args several times,
+--- (to the same or different peers), this method returns the
+--- invariant part, so it can be cached to reduce encoding overhead
+function wrpc_service:encode_args(name, ...)
+  local rpc = self:get_method(name)
+  if not rpc then
+    return nil, string.format("no method %q", name)
+  end
+
+  local num_args = select('#', ...)
+  local payloads = table.new(num_args, 0)
+  for i = 1, num_args do
+    payloads[i] = assert(pb.encode(rpc.input_type, select(i, ...)))
+  end
+
+  return rpc, payloads
+end
+
+
+local wrpc_peer = {
+  encode = pb.encode,
+  decode = pb.decode,
+}
+wrpc_peer.__index = wrpc_peer
+
+--- a `peer` object holds a (websocket) connection and a service.
+function wrpc.new_peer(conn, service, opts)
+  opts = opts or {}
+  return setmetatable(merge({
+    conn = conn,
+    service = service,
+    seq = 1,
+    request_queue = (conn and not conn.close) and Queue.new(),
+    response_queue = {},
+    closing = false,
+    channel_dict = opts.channel and ngx.shared[CHANNEL_DICT_NAME],
+    _receiving_thread = nil,
+  }, opts), wrpc_peer)
+end
+
+
+function wrpc_peer:close()
+  self.closing = true
+  self.conn:send_close()
+  if self.conn.close then
+    self.conn:close()
+  end
+end
+
+
+function wrpc_peer:send(d)
+  if self.request_queue then
+    return self.request_queue:push(d)
+  end
+
+  return self.conn:send_binary(d)
+end
+
+function wrpc_peer:receive()
+  while true do
+    local data, typ, err = self.conn:recv_frame()
+    if not data then
+      return nil, err
+    end
+
+    if typ == "binary" then
+      return data
+    end
+
+    if typ == "close" then
+      kong.log.notice("Received WebSocket \"close\" frame from peer")
+      return self:close()
+    end
+  end
+end
+
+--- RPC call.
+--- returns the call sequence number, doesn't wait for response.
+function wrpc_peer:call(name, ...)
+  local rpc, payloads = assert(self.service:encode_args(name, ...))
+  return self:send_encoded_call(rpc, payloads)
+end
+
+
+function wrpc_peer:call_wait(name, ...)
+  local waiter = semaphore_waiter()
+
+  local seq = self.seq
+  self.response_queue[seq] = waiter
+  local new_seq = self:call(name, ...)
+  assert(new_seq == seq)
+
+  waiter.smph:wait()
+  return waiter.data
+end
+
+
+--- Part of wrpc_peer:call()
+--- This performs the per-call parts.  The arguments
+--- are the return values from wrpc_peer:encode_args(),
+--- either directly or cached (to repeat the same call
+--- several times).
+function wrpc_peer:send_encoded_call(rpc, payloads)
+  self:send_payload({
+    mtype = "MESSAGE_TYPE_RPC",
+    svc_id = rpc.service_id,
+    rpc_id = rpc.rpc_id,
+    payload_encoding = "ENCODING_PROTO3",
+    payloads = payloads,
+  })
+  return self.seq
+end
+
+--- little helper to ease grabbing an unspecified number
+--- of values after an `ok` flag
+local function ok_wrapper(ok, ...)
+  return ok, {n = select('#', ...), ...}
+end
```

Comment:
:open_mouth: This is magic!!!

Reply:
i don't like it too much, it's there to preserve the plurality of the `payloads` argument, both as input and output